virtual_rainforest/core/data.py

```diff
@@ -0,0 +1,686 @@
+
+    # Now ensure sorting and any subsetting:
+    # https://stackoverflow.com/questions/8251541
+    da_sortorder = np.argsort(da_cell_ids)
+    gridid_pos = np.searchsorted(da_cell_ids[da_sortorder], data.grid.cell_id)
+    da_indices = da_sortorder[gridid_pos]
+
+    return darray.isel(cell_id=da_indices)
+
+
+@register_axis_validator("spatial", (("cell_id",), (), ("__any__",)))
+def spld_cellid_dim_any(data: Data, darray: DataArray) -> DataArray:
+    """Spatial loader for cell id dimension onto any grid.
+
+    In this loader, the DataArray only has a cell_id dimension so assumes that the
+    values are provided in the same sequence as the grid cell ids. Because this method
+    simply maps data to grid cells by id, it should applies to _any_ arbitrary grid
+    setup.
+    """
+
+    # Cell ID is only a dimenson with a give length - assume the order correct
+    # and check the right number of cells found
+    n_found = darray["cell_id"].size
+    if data.grid.n_cells != n_found:
+        raise ValueError(
+            f"Grid defines {data.grid.n_cells} cells, data provides {n_found}"
+        )
+
+    return darray
+
+
+@register_axis_validator("spatial", (("x", "y"), ("x", "y"), ("square",)))
+def spld_xy_coord_square(data: Data, darray: DataArray) -> DataArray:
+    """Spatial loader for XY coordinates onto a square grid."""
+
+    # Get x and y coords to check the extents and cell coverage.
+    #
+    # TODO - Note that mapping all the cells here is a bit extreme with a square grid -
+    # could just map one row and column to confirm the indexing into the data, but this
+    # is a more general solution.
+
+    # Use .stack() to convert the axis into stacked pairs of values
+    xy_pairs = darray.stack(cell=("y", "x")).coords
+    idx_pairs = darray.drop_vars(("x", "y")).stack(cell=("y", "x")).coords
+
+    # Get the mapping of points onto the grid
+    idx_x, idx_y = data.grid.map_xy_to_cell_indexing(
+        x_coords=xy_pairs["x"].values,
+        y_coords=xy_pairs["y"].values,
+        x_idx=idx_pairs["x"].values,
+        y_idx=idx_pairs["y"].values,
+        strict=False,
+    )
+
+    # TODO - fine a way to enable strict = True - probably just kwargs.
+
+    # Now remap the grids from xy to cell_id - this uses the rather under
+    # described vectorized indexing feature in xarray:
+    #
+    #   https://docs.xarray.dev/en/stable/user-guide/indexing.html
+    #
+    # Specifically using DataArray.isel to select indexed dimensions by name, which
+    # avoids issues with different permutations of the axes - and map XY onto a common
+    # cell_id
+
+    darray = darray.isel(
+        x=DataArray(idx_x, dims=["cell_id"]), y=DataArray(idx_y, dims=["cell_id"])
+    )
+
+    return darray
+
+
+@register_axis_validator("spatial", (("x", "y"), (), ("square",)))
+def spld_xy_dim_square(data: Data, darray: DataArray) -> DataArray:
+    """Spatial loader for XY dimensions onto a square grid."""
```

Comment:
Probably useful for the docstring here to explain the differences from the function above (which currently has an identical docstring)

Reply:
It's not quite identical - dimensions vs coordinates - but yes these docstrings are not complete. Not sure how they passed to be honest - I thought `flake8` would catch that.